### ISSUE TITLE
How to set up redirect URIs?

### ISSUE THREAD

**Issue by Chaphasilor** [XDGFX/ultrasonics-api#1]
I can't find documentation on this and my attempts so far didn't work out.  

I've set up the API server on my server, accessible on port X. I get the default response when navigating to `/api` as well, so that seems to work as expected.

I've configured applications for both Spotify and Deezer on their respective dashboards and used the IDs and keys for the necessary environment variables.

Now I'm stuck at the redirect URIs. I've configured the Spotify one to <local server IP address>:X/api/spotify/auth, which I took [from here](https://github.com/XDGFX/ultrasonics-api/blob/master/ultrasonics_api/services/spotify.py#L125-L155).  
In that function it seems like the redirect URI is hard-coded to the heroku instance, although I could be wrong.  
I did the same for deezer, and this time there was no hard-coded backlink in [the auth function](https://github.com/XDGFX/ultrasonics-api/blob/master/ultrasonics_api/services/deezer.py#L92-L121).

Now, when I go to ultrasonics and try to authenticate with Spotify or Deezer, I always end up with an error saying "INVALID_CLIENT: Invalid redirect URI" (for Spotify) or "You must specify a valid redirect URI" (for Deezer).  
But I have a redirect URI set up on the developer dashboard for both applications.  
Are IPs not allowed for these redirect URIs? Given that the client is being redirected and has access to that local IP, it should technically work, right?

Some more documentation on the matter would be really helpful if we are to migrate away form the heroku-hosted API :)

**Edit:** Looking at the URL that is used to send the authentication request (to Spotify's/Deezer's website), it seems like the heroku instance URI is hard-coded for both services. Could this be changed please, for example by adding an environment variable for it?

**Comment by Chaphasilor** [comment]
These would be the places where the environment variable is needed:  

Deezer:

https://github.com/XDGFX/ultrasonics-api/blob/d1994df659193d08158bea5f6eba0d19f556b8d2/ultrasonics_api/services/deezer.py#L75

Spotify:

https://github.com/XDGFX/ultrasonics-api/blob/d1994df659193d08158bea5f6eba0d19f556b8d2/ultrasonics_api/services/spotify.py#L92  

https://github.com/XDGFX/ultrasonics-api/blob/d1994df659193d08158bea5f6eba0d19f556b8d2/ultrasonics_api/services/spotify.py#L150

**Comment by Chaphasilor** [comment]
I've managed to patch the python files to load a base URL for the API from an environment variable and use that for the redirect URIs. Spotify authentication is now working, but for Deezer I'm getting the following error:

```
Access Denied
You don't have permission to access "http://connect.deezer.com/oauth/auth.php?" on this server.
```

My request looks like this:  
`https://connect.deezer.com/oauth/auth.php?app_id=<redacted>&redirect_uri=http://<IP>:<X>/api/deezer/auth&perms=basic_access,offline_access,manage_library,delete_library`, which aligns with the documentation at https://developers.deezer.com/api/oauth. Any idea what I'm missing here?

Once I manage to get this working I'd be happy to create a PR for this :)

**Comment by Chaphasilor** [comment]
Okay, it seems like Deezer's `Application domain` doesn't support plain IP addresses, whereas Spotify has no problems with that.  
So in order to authenticate with Deezer, you need to either use an actual domain for the API server, or use mDNS in order to replace you IP with an equivalent domain name...

**Edit:** it seems like mDNS resolution does not work from the ultrasonics python server / docker image, so **you'll have to use the actual IP there (in the ultrasonics settings on the website)**, otherwise the initial auth will work, but renewing tokens will not, because for that the ultrasonics server has to make a request to the ultrasonics-api server :)

**Comment by MauliQT** [comment]
Hi i am trying to setup the api rn, but i cant access the spotify auth site and I don't know how to do it, and you said you got it running, could you please help me out

**Comment by Chaphasilor** [comment]
1. Go here and log in with your Spotify Account: https://developer.spotify.com/dashboard/
2. Create a new application, name it e.g. "ultrasonics"
3. In your app's settings, set the redirect URL to the URL where your ultrasonics-api server is reachable

This might not be enough to help you out, if you still have problems I can send you some screenshots once I get back to my PC :)

**Comment by MauliQT** [comment]
Does it have to be a public ip or can i do it just with my local ip. And if not is it possible to reverse proxy the api through nginx
Edit: tried routing it trough nginx it works, but still cant auth spotify 
<img width="599" alt="Screenshot 2022-11-07 at 16 36 12" src="https://user-images.githubusercontent.com/87561264/200350756-9f5669fa-e98b-48dc-ae09-65d1b286fb47.png">
<img width="1306" alt="Screenshot 2022-11-07 at 16 37 03" src="https://user-images.githubusercontent.com/87561264/200350938-94bac701-ff8b-4e35-addc-b1931dd37da0.png">
<img width="444" alt="Screenshot 2022-11-07 at 17 06 19" src="https://user-images.githubusercontent.com/87561264/200357719-2b1a4a8c-8bef-4764-90b3-0b601c31cc4a.png">

Edit: changed Domain from music.fjcx764.uk to music.fjcx764.uk/api

**Comment by Chaphasilor** [comment]
Yeah local IPs should work for Spotify. What you seem to be missing is the correct path for the redirect URL, it should be http://192.168.2.124:8003/api/auth/spotify and not just http://192.168.2.124:8003 :)

**Comment by MauliQT** [comment]
oh with the auth/spotify aswell

**Comment by MauliQT** [comment]
still doesn't work, could you do some screenshots of your setup

**Comment by MauliQT** [comment]
Nvm works now just had to restart everything

**Comment by Chaphasilor** [comment]
Great!

**Comment by MauliQT** [comment]
anyways now im trying to get Deezer to work, im using my domain now it works with spotify but deezer doesnt want to work. My question is what goes where and does it work for you. I'm using a fork where i can edit the redirect uris inside the docker container (https://github.com/Frazew/ultrasonics-api) 
<img width="469" alt="Screenshot 2022-11-09 at 16 48 46" src="https://user-images.githubusercontent.com/87561264/200876358-cf92a60f-6e7b-4a80-a682-6b64a38f9d1b.png">
btw thank you for your help so far

**Comment by haudankaivajasi** [comment]
Okay, I'm just baffled now as I'm trying to get the self hosted ultrasonics-api to work on docker. I'm running both on my server and used the given docker commands to run the containers with api keys and secrets. 

I have entered ultrasonics-api URL as: http://server_ip:8003/ in the Ultrasonics settings screen and added http://server_ip:8003/spotify/auth/request to Spotify api Redirect URIs. I also tried https://ultrasonics-api.mydomain.com/ in Ultrasonics settings and in https://ultrasonics-api.mydomain.com/spotify/auth/request in Spotify. I've restarted both Ultrasonics and the api container after changes just to be sure.

Every time I try to authenticate with Spotify it gives me a "Not found" page.

Is there a problem with the api container or what might cause this behavior?

**Comment by haudankaivajasi** [comment]
> Okay, I'm just baffled now as I'm trying to get the self hosted ultrasonics-api to work on docker. I'm running both on my server and used the given docker commands to run the containers with api keys and secrets.
> 
> I have entered ultrasonics-api URL as: http://server_ip:8003/ in the Ultrasonics settings screen and added http://server_ip:8003/spotify/auth/request to Spotify api Redirect URIs. I also tried https://ultrasonics-api.mydomain.com/ in Ultrasonics settings and in https://ultrasonics-api.mydomain.com/spotify/auth/request in Spotify. I've restarted both Ultrasonics and the api container after changes just to be sure.
> 
> Every time I try to authenticate with Spotify it gives me a "Not found" page.
> 
> Is there a problem with the api container or what might cause this behavior?

Ok, I got it to work.
First I had to add [this](https://github.com/MauliQT/ultrasonics-api/blob/0805b07d4fedbed23e202932ea20d89079a0d83f/ultrasonics_api/services/spotify.py) modified spotify.py file to the container with `-v /file/location//ultrasonics/api/spotify.py:/app/ultrasonics_api/services/spotify.py` and also add `-e SPOTIFY_REDIRECT_URI=http://my_internal_ip:8003/api/spotify/auth \` to the docker command when running the container. This way it won't redirect straight to heroku app.

Then I added `http://my_internal_ip:8003/api/` to ultrasonics settings via the web interface. Lastly I added `http://my_internal_ip:8003/api/spotify/auth` to the Spotify app's "Redirect URIs" section.

I hope this helps someone struggling to get this to work!

**Comment by bartvaneck** [comment]
Is there a possibility to get the modified spotify.py again?
